sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala

```diff
@@ -64,7 +68,17 @@ case class PythonUDF(
 
   override def toString: String = s"$name(${children.mkString(", ")})#${resultId.id}$typeSuffix"
 
-  final override val nodePatterns: Seq[TreePattern] = Seq(PYTHON_UDF)
+  // SPARK-41633: We should check whether to update the node patterns when adding a new eval type.
+  private def nodePatternsOfPythonFunction: Seq[TreePattern] = {
+    if (PythonUDF.isGroupedAggPandasUDFEvalType(evalType)) {
+      Seq(AGGREGATE_EXPRESSION)
```

Comment:
I think it's better to make node pattern match the node type. We created node patterns to speed up `isInstanceOf` check, not to replace it. Even if we fix the node pattern here, `expr.exists(_.isInstanceOf[AggregateExpression])` is still broken.

I think we should have a `PythonUDAF` that indicates aggregate function explicitly, instead of relying on node pattern.

Reply:
> Even if we fix the node pattern here, expr.exists(_.isInstanceOf[AggregateExpression]) is still broken.

I am not getting your point about this one. Can you explain it?

Reply:
@rxin Given Wenchen is against the idea and there is little we can do for future-proofing, do you mind matching the aggregation PythonUDF in https://github.com/apache/spark/pull/39134 instead of adding this node pattern?

Reply:
I agree that we should have a subtype for aggregate. That's a much safer change. Should've done that from the beginning IMO.